javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -783,7 +824,51 @@ public boolean isTraversableAgain() {
 
     @Override
     public Iterator<T> iterator() {
-        return isEmpty() ? Iterator.empty() : trie.iterator();
+        return isEmpty() ? Iterator.empty()
+                         : new VectorIterator();
+    }
+
```

Comment:
I know this sounds weird but it is the only explanation. Creating the Iterators is constant effort. Their implementation is as efficient as the solution with one Iterator. Beside Iterator creation there could be only one more reason - the concat Iterator is implemented inefficient. But it is only a simple iteration and switch of existing Iterators. That can't be the bottleneck.

We should investigate it. It is a good opportunity to get my hands dirty regarding the benchmarks.